xen/arch/arm/coproc/coproc.h

```diff
@@ -29,28 +29,124 @@
 
 #include "schedule.h"
 
-/* coproc memory range */
-struct mmio {
+/*
+ * In fields and function names following prefixes mean:
+ * - [p] Platform specific. Means coprocessor IP specific stuff.
+ * - [m] Machine specific. Means some stuff specific to a SoC which integrates
+ *       a coprocessor.
+ * - [v] Virtual. Means a virtual coprocessor related stuff.
+ */
+
+/*
+ * Coprocessor platform memory mapped IO range description
+ *
+ * This is a structure what describes one particular mmio range of a
+ * coprocessor IP block.
+ * This description expected to be common for different SoCs which could
+ * integrate this particular coprocessor IP.
+ * This structure is used as static with following fields initialized at the
+ * compilation time:
+ *     name - to find in DT, ignored if a coproc has only one mmio range
+ *     size - >0 for sanity check, =0 - no sanity check needed, filled from DT
+ *               (f.e. for SRAM bank)
+ *     ops  - IO registers access emulation (r/w) handlers
+ */
+struct pcoproc_mmio {
+    char *name;
+    u64 size;
+    struct mmio_handler_ops *ops;
+};
+
+/*
+ * Coprocessor platform IRQ description
+ *
+ * This is a structure what describes one particular IRQ of a coprocessor IP
+ * block. This description is expected to be common for different SoCs which
+ * could integrate this particular coprocessor IP.
+ * This structure is used as static with following fields initialized at the
+ * compilation time:
+ *     name - to find in DT, ignored if a coproc has only one IRQ
+ *     handler - irq handler function aware of this particular IRQ functionality
+ */
+struct pcoproc_irq {
+    char *name;
+    void (*handler)(int, void *, struct cpu_user_regs *);
+};
+
+/*
+ * Coprocessor platform description
+ * A structure what gathers coproc description (irq and mmio so far)
+ */
+struct pcoproc_desc {
+    u64 p_mmio_num;
+    struct pcoproc_mmio *p_mmio;
+    u64 p_irq_num;
+    struct pcoproc_irq *p_irq;
+};
+
+/*
+ * Coprocessor machine IO memory range
+ * This structure is initialized on dt parsing for each io range found for a
+ * coprocessor
+ */
+struct mcoproc_mmio {
+    /* iorange machine base address, taken from a device-tree by name */
     u64 addr;
+    /* mmio range size from dt */
     u64 size;
-    /* ioremapped addr */
+    /* ioremapped base address, mapped on coproc iniitalization */
     void __iomem *base;
+    /* pointer to a correspondent platform mmio range description */
+    struct pcoproc_mmio *p_mmio;
+};
 
-    struct coproc_device *coproc;
+/*
+ * Coprocessor machine IRQ
+ * This structure is initialized on dt parsing for each irq found for a
+ * coprocessor.
+ * An interrupt handler receives a triggered irq number and this seems to be
+ * enough to inject an interrupt to a domain.
+ * This description will be used in cases when we need to inject an interrupt
+ * manually (without receiving it from HW).
+ */
+struct mcoproc_irq {
+    /* actual IRQ number */
+    unsigned int irq; 
```

Comment:
I am a bit confused. It is an irq that will be requested and passed to irq_handler? Shouldn't it be just "int"?